indexing-service/src/main/java/org/apache/druid/indexing/common/task/AbstractBatchIndexTask.java

```diff
@@ -443,6 +466,42 @@ protected boolean tryTimeChunkLock(TaskActionClient client, List<Interval> inter
     return true;
   }
 
+  private TaskLockHelper createLockHelper(LockGranularity lockGranularity)
+  {
+    return new TaskLockHelper(
+        lockGranularity == LockGranularity.SEGMENT,
+        determineLockType(lockGranularity)
+    );
+  }
+
+  /**
+   * Determines the type of lock to use with the given lock granularity.
+   */
+  private TaskLockType determineLockType(LockGranularity lockGranularity)
+  {
+    if (lockGranularity == LockGranularity.SEGMENT) {
+      return TaskLockType.EXCLUSIVE;
+    }
+
+    final String contextLockType = getContextValue(Tasks.TASK_LOCK_TYPE);
+    final TaskLockType lockType;
+    if (contextLockType == null) {
+      lockType = getContextValue(Tasks.USE_SHARED_LOCK, false)
+                 ? TaskLockType.SHARED : TaskLockType.EXCLUSIVE;
+    } else {
+      lockType = TaskLockType.valueOf(contextLockType);
+    }
+
+    final IngestionMode ingestionMode = getIngestionMode();
+    if ((lockType == TaskLockType.SHARED || lockType == TaskLockType.APPEND)
+        && ingestionMode != IngestionMode.APPEND) {
+      // Lock types SHARED and APPEND are allowed only in APPEND ingestion mode
+      return Tasks.DEFAULT_TASK_LOCK_TYPE;
+    } else {
+      return lockType;
```

Comment:
should we instead throw an exception when task lock type is Append but ingestion mode is not append? That seems like something that would only happen because of a mistake.

Reply:
Yeah, makes sense. And I suppose a validation should be done at the time of submitting the ingestion spec as well so that an incorrectly configured task fails fast.